pkg/cmd/task/task_download.go

```diff
@@ -0,0 +1,174 @@
+		RunE: func(cmd *cobra.Command, args []string) error {
+			task, err := findTask(ctx, args[0])
+			if err != nil {
+				return err
+			}
+
+			lastStatus := task.Statuses[0]
+			containerID := lastStatus.ContainerStatus.GetContainerID().Value
+
+			executorID := task.TaskID.Value
+			if task.ExecutorID != nil {
+				executorID = (*task.ExecutorID).Value
+			}
+
+			c := mesos.NewClient(pluginutil.HTTPClient(""))
+			paths, err := c.Debug(task.AgentID.Value)
+			if err != nil {
+				return err
+			}
+
+			path, err := getPath(paths, task.FrameworkID.Value, executorID, containerID)
+			if err != nil {
+				return err
+			}
+
+			pattern := "/"
+			if len(args) == 2 {
+				pattern = args[1]
+			}
+
+			return downloadDir(c, ctx, task.AgentID.Value, path, targetDir, pattern)
+		},
+	}
+
+	dir, err := os.Getwd()
+	if err == nil {
+		targetDir = dir
+	}
+	cmd.Flags().StringVar(&targetDir, "target-dir", targetDir, "Target directory of the download. Defaults to $PWD")
+	return cmd
+}
+
+func getPath(paths map[string]string, framework, executor, container string) (string, error) {
+	for k, v := range paths {
+		if strings.Contains(k, "/frameworks/"+framework+"/executors/"+executor+"/runs/"+container) {
+			return v, nil
+		}
+	}
+	return "", fmt.Errorf("unable to find task")
+}
+
+func downloadDir(c *mesos.Client, ctx api.Context, agentID, agentPath, targetDir, pattern string) error {
+	if err := ctx.Fs().MkdirAll(targetDir, 0755); err != nil {
+		return err
+	}
+
+	parent := parentPath(agentPath, pattern)
+	base := basePath(pattern)
+	files, err := c.Browse(agentID, parent)
+	if err != nil {
+		return err
+	}
+
+	errors := make(chan error, len(files))
+	for _, file := range files {
+		matched, err := matchFile(base, path.Base(file.Path))
+		if err != nil {
+			return fmt.Errorf("'%s' as pattern not supported: %v", pattern, err)
+		}
+
+		if !matched {
+			errors <- nil
+			continue
+		}
+
+		go func(f mesos.File) {
+			if strings.HasPrefix(f.Mode, "d") {
+				err = downloadDir(c, ctx, agentID, f.Path, filepath.Join(targetDir, path.Base(f.Path)), "/")
+				if err != nil {
+					errors <- err
+					return
+				}
+			} else {
+				b, err := c.Download(agentID, f.Path)
+				if err != nil {
+					errors <- fmt.Errorf("could not fetch file '%s': %v", path.Base(f.Path), err)
+					return
+				}
+
+				out, err := ctx.Fs().Create(filepath.Join(targetDir, path.Base(f.Path)))
+				if err != nil {
+					errors <- fmt.Errorf("unable to create file '%s'", path.Base(f.Path))
+					return
+				}
+				defer out.Close()
+				_, err = out.Write(b)
+				if err != nil {
+					errors <- fmt.Errorf("unable to write to file '%s'", path.Base(f.Path))
+					return
+				}
+			}
+			errors <- nil
+		}(file)
+	}
+
+	var failed []string
+	for _, file := range files {
+		err := <-errors
+		if err != nil {
+			failed = append(failed, path.Base(file.Path))
+		}
+	}
+	if len(failed) != 0 {
+		return fmt.Errorf("Download of following files failed: [%s]", strings.Join(failed, ", "))
```

Comment:
I don't think `%s` needs to be surrounded by square brackets in that error message.

Reply:
This doesn't really work as I want anyway because when I pull an error from the channel it doesn't necessarily belong to the file I am looping over at that moment. Will change it to something different.